apollo-router/src/introspection.rs

```diff
@@ -54,10 +54,16 @@ impl IntrospectionCache {
     ) -> ControlFlow<graphql::Response, ()> {
         Self::maybe_lone_root_typename(schema, doc)?;
         if doc.operation.is_query() {
-            if doc.has_explicit_root_fields && doc.has_schema_introspection {
-                ControlFlow::Break(Self::mixed_fields_error())?;
+            if doc.has_schema_introspection {
+                if doc.has_explicit_root_fields {
+                    ControlFlow::Break(Self::mixed_fields_error())?;
+                } else {
+                    ControlFlow::Break(self.cached_introspection(schema, key, doc).await)?
+                }
             } else if !doc.has_explicit_root_fields {
-                ControlFlow::Break(self.cached_introspection(schema, key, doc).await)?
+                // root __typename only, probably a small query
+                // Execute it without caching:
+                ControlFlow::Break(Self::execute_introspection(schema, doc))?
```

Comment:
Shouldn't this be handled by the `maybe_lone_root_typename` call earlier in the function? Or maybe we can remove that function (which doesn't handle fragments), and only rely on this branch now?

Reply:
`maybe_lone_root_typename` handles exactly `{__typename}` (which is very common) and ~nothing else. This handles cases where `__typename` is the only field, but fragments or directives may be used.

It’s true that the `maybe_lone_root_typename` fast path becomes less impactful with this PR since the cache is bypassed either way. I wouldn’t oppose remove it, though I don’t know if this fast path is still faster than going through the executor machinery.